### PR TITLE
Resolved potential issues with the distance measurements

### LINKED ISSUE #417: Reroute never appears if currentstep has only one geometry point
Reroute never appears if current step has only one geometry point.

`https://github.com/mapbox/mapbox-navigation-android/blob/79c9c4e023e04e8076d980872de7383c353ed89b/libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/offroute/OffRouteDetector.java#L119`

TurfMisc.pointOnLine returns a point with infinity coordinates {type='Point', coordinates=Position [longitude=Infinity, latitude=Infinity, altitude=NaN]}

It leads to a NaN value for the computation of the distance. --> Reroute not detected

You could compute the distance between futurPosition and the geometry point in case of only one geometry point exists.

### PR DESCRIPTION
**Pull request by cammace**
Closes #417

This resolves a few issues when measuring the users true distance to the closest point on the route line.
- If user location falls directly on route line, the distance is 0
- if step geometry coordinates only has one point, we measure the distance from that point to the raw location (Before, this was causing `NaN`)

I've also pulled out the static method for easier testing, something we should probably start doing more.

### CODE REVIEW

File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/utils/MeasurementUtils.java [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+      return 0;
+    }
+
+    // Get the lineString from the step geometry.
+    LineString lineString = LineString.fromPolyline(step.geometry(), Constants.PRECISION_6);
+
+    // Make sure that the step coordinates isn't empty. If the points equal each other, the distance
+    // is obviously zero, so return 0.
+    if (lineString.coordinates().isEmpty()
+      || usersRawLocation.equals(lineString.coordinates().get(0))) {
```
Comment (1ec5) [inline libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/utils/MeasurementUtils.java]: Shouldn’t `TurfMeasurement.distance()` be responsible for checking whether the two coordinates are equal? Or does this get around a performance hit from calling `TurfMisc.pointOnLine()` unnecessarily?

File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/utils/MeasurementUtils.java [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+      return 0;
+    }
+
+    Feature feature = TurfMisc.pointOnLine(usersRawLocation, lineString.coordinates());
```
Comment (palto-ccadoret) [inline libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/utils/MeasurementUtils.java]: Should not call pointOnLine if lineString contains only one coordinate.
Why not return TurfMeasurement.distance(usersRawLocation, lineString.coordinates().get(0), UNIT_METERS) directly ?

**Comment by danesfeder** [comment]
@cammace is this good to move?  Or do the comments still need to be addressed

**Comment by cammace** [comment]
I agree with @1ec5 that the size of the list should be checked inside turf, I've opened a ticket for fixing this: https://github.com/mapbox/mapbox-java/issues/632

File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/utils/MeasurementUtils.java (hunk 1 above)
Comment (cammace) [reply libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/utils/MeasurementUtils.java]: Looking deeper into this, Checking the two Point objects for equality does, in fact, prevent a performance hit. If we removed this check, the end result will be the same, a distance of 0 with no runtime exceptions being thrown. [This test](https://github.com/mapbox/mapbox-java/blob/cbec9e10d98e6676bbfcfc71a69462e5ff5007ff/services-turf/src/test/java/com/mapbox/turf/TurfMiscTest.java#L139-L151) confirms this.

**Pull request merged**